File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Random;

class DnDCharacter {
    Random rand;
    private final int strength;
    private final int dexterity;
    private final int constitution;
    private final int intelligence;
    private final int wisdom;
    private final int charisma;

    public DnDCharacter() {
        rand = new Random();
        strength = ability();
        dexterity = ability();
        constitution = ability();
        intelligence = ability();
        wisdom = ability();
        charisma = ability();
    }

    int ability() {
        return rand.ints(4, 1, 7)
                .sorted().skip(1).sum();
    }

    public int getStrength() {
        return strength;
    }

    public int getDexterity() {
        return dexterity;
    }

    public int getConstitution() {
        return constitution;
    }

    public int getIntelligence() {
        return intelligence;
    }

    public int getWisdom() {
        return wisdom;
    }

    public int getCharisma() {
        return charisma;
    }

    int modifier(int input) {
        return Math.floorDiv(input - 10, 2);
    }

    public int getHitpoints() {
        return 10 + modifier(constitution);
    }
}

File: php/dnd-character/DndCharacter.php
<?php
declare(strict_types=1);
class DndCharacter
{
    private const ABILITIES = [
        'strength',
        'dexterity',
        'constitution',
        'intelligence',
        'wisdom',
        'charisma'
    ];
    public int $ability;
    public int $strength;
    public int $dexterity;
    public int $constitution;
    public int $intelligence;
    public int $wisdom;
    public int $charisma;
    private const HITPOINTS = 10;
    private const THROWS = 3;
    private const MIN_SUM_DICES = 1 * self::THROWS;
    private const MAX_SUM_DICES = 6 * self::THROWS;
    public static function ability(): int
    {
        return self::throwAndSumDices();
    }
    public static function generate(): stdClass
    {
        $character = new stdClass();
        foreach (self::ABILITIES as $ability) {
            $character->{$ability} = self::throwAndSumDices();
        }
        $character->hitpoints = self::HITPOINTS + self::modifier($character->constitution);
        return $character;
    }
    public static function modifier(int $input): int
    {
        return (int) floor(($input - self::HITPOINTS) / 2);
    }
    private static function throwAndSumDices(): int
    {
        return random_int(self::MIN_SUM_DICES, self::MAX_SUM_DICES);
    }
}

